includes/import/abstract-wc-product-importer.php

```diff
@@ -0,0 +1,941 @@
+	}
+
+	/**
+	 * Process a single item and save.
+	 *
+	 * @param  array $data Raw CSV data.
+	 * @return WC_Product|WC_Error
+	 */
+	protected function process_item( $data ) {
+		// Ignore IDs and create new products.
+		// @todo Mike said that we should have something to force create.
+		$force_create = false;
+
+		try {
+			$object = $this->prepare_product( $data, $force_create );
+
+			if ( is_wp_error( $object ) ) {
+				return $object;
+			}
+
+			$object->save();
+
+			// Clean cache for updated products.
+			$this->clear_cache( $object );
+
+			return $object->get_id();
+		} catch ( WC_Data_Exception $e ) {
+			return new WP_Error( $e->getErrorCode(), $e->getMessage(), $e->getErrorData() );
+		} catch ( Exception $e ) {
+			return new WP_Error( 'woocommerce_product_csv_importer_error', $e->getMessage(), array( 'status' => $e->getCode() ) );
+		}
+	}
+
+	/**
+	 * Clear product cache.
+	 *
+	 * @param WC_Product $object Product instance.
+	 */
+	protected function clear_cache( $object ) {
+		$id = $object->get_id();
+
+		if ( 'variation' === $object->get_type() ) {
+			$id = $object->get_parent_id();
+		}
+
+		wc_delete_product_transients( $id );
+		wp_cache_delete( 'product-' . $id, 'products' );
+	}
+
+	/**
+	 * Prepare a single product for create or update.
+	 *
+	 * @param  array $data     Row data.
+	 * @param  bool  $creating If should force create a new product.
+	 * @return WC_Product|WP_Error
+	 */
+	protected function prepare_product( $data, $force_create = false ) {
+		$id = ! $force_create && isset( $data['id'] ) ? absint( $data['id'] ) : 0;
+
+		// Type is the most important part here because we need to be using the correct class and methods.
+		if ( isset( $data['type'] ) ) {
+			if ( ! in_array( $data['type'], array_keys( wc_get_product_types() ), true ) ) {
+				return new WP_Error( 'woocommerce_product_csv_importer_invalid_type', __( 'Invalid product type.', 'woocommerce' ), array( 'status' => 401 ) );
+			}
+
+			$classname = WC_Product_Factory::get_classname_from_product_type( $data['type'] );
+
+			if ( ! class_exists( $classname ) ) {
+				$classname = 'WC_Product_Simple';
+			}
+
+			$product = new $classname( $id );
+		} elseif ( isset( $data['id'] ) ) {
+			$product = wc_get_product( $id );
+		} else {
+			$product = new WC_Product_Simple();
+		}
+
+		// @todo need to check first how we'll handle variations.
+		if ( 'variation' === $product->get_type() ) {
+			$product = $this->save_variation_data( $product, $data );
+		} else {
+			$product = $this->save_product_data( $product, $data );
+		}
+
+		return apply_filters( 'woocommerce_product_csv_import_pre_insert_product_object', $product, $data );
+	}
+
+	/**
+	 * Set product data.
+	 *
+	 * @param WC_Product $product Product instance.
+	 * @param array      $data    Row data.
+	 *
+	 * @return WC_Product
+	 */
+	protected function save_product_data( $product, $data ) {
```

Comment:
Most of these `$data` keys map 1:1 with the actual product schema - why not just use set_props?

Prepare $data if needed, pass to set_props, done?